library/core/src/iter/adapters/intersperse.rs

```diff
@@ -0,0 +1,101 @@
+use crate::ops::Try;
+
+use super::Peekable;
+
+/// An iterator adapter that places a separator between all elements.
+#[unstable(feature = "iter_intersperse", reason = "recently added", issue = "none")]
+#[derive(Debug, Clone)]
+pub struct Intersperse<I: Iterator>
+where
+    I::Item: Clone,
+{
+    separator: I::Item,
+    iter: Peekable<I>,
+    needs_sep: bool,
+}
+
+impl<I: Iterator> Intersperse<I>
+where
+    I::Item: Clone,
+{
+    pub(in crate::iter) fn new(iter: I, separator: I::Item) -> Self {
+        Self { iter: iter.peekable(), separator, needs_sep: false }
+    }
+}
+
+#[unstable(feature = "iter_intersperse", reason = "recently added", issue = "none")]
+impl<I> Iterator for Intersperse<I>
+where
+    I: Iterator,
+    I::Item: Clone,
+{
+    type Item = I::Item;
+
+    #[inline]
+    fn next(&mut self) -> Option<I::Item> {
+        if self.needs_sep && self.iter.peek().is_some() {
+            self.needs_sep = false;
+            Some(self.separator.clone())
+        } else {
+            self.needs_sep = true;
+            self.iter.next()
+        }
+    }
+
+    fn fold<B, F>(mut self, init: B, mut f: F) -> B
+    where
+        Self: Sized,
+        F: FnMut(B, Self::Item) -> B,
+    {
+        let mut accum = init;
+
+        // Use `peek()` first to avoid calling `next()` on an empty iterator.
+        if !self.needs_sep || self.iter.peek().is_some() {
+            if let Some(x) = self.iter.next() {
+                accum = f(accum, x);
+            }
+        }
+
+        let element = &self.separator;
+
+        self.iter.fold(accum, |mut accum, x| {
+            accum = f(accum, element.clone());
+            accum = f(accum, x);
+            accum
+        })
+    }
+
+    fn try_fold<B, F, R>(&mut self, init: B, mut f: F) -> R
+    where
+        Self: Sized,
+        F: FnMut(B, Self::Item) -> R,
+        R: Try<Ok = B>,
+    {
+        let mut accum = init;
+
+        // Use `peek()` first to avoid calling `next()` on an empty iterator.
+        if !self.needs_sep || self.iter.peek().is_some() {
+            if let Some(x) = self.iter.next() {
+                accum = f(accum, x)?;
+            }
+        }
+
+        let element = &self.separator;
+
+        self.iter.try_fold(accum, |mut accum, x| {
+            accum = f(accum, element.clone())?;
+            accum = f(accum, x)?;
+            Try::from_ok(accum)
+        })
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let (lo, hi) = self.iter.size_hint();
+        let next_is_elem = !self.needs_sep;
+        let lo = lo.saturating_add(lo).saturating_sub(next_is_elem as usize);
+        let hi = hi
+            .and_then(|hi| hi.checked_add(hi))
+            .and_then(|hi| hi.checked_sub(next_is_elem as usize));
```

Comment:
Why shouldn't it subtract first and then add? This could allow one more value on the upper bound if it hit `usize::MAX`. Will `lo` and `hi` every be `0`?
```suggestion
        let lo = lo.saturating_sub(next_is_elem as usize).saturating_add(lo);
        let hi = hi
            .and_then(|hi| hi.checked_sub(next_is_elem as usize))
            .and_then(|hi| hi.checked_add(hi));
```

Reply:
For `hi`, in addition to doing the subtraction first, it should use `saturating_sub` not `checked_sub` because if `hi` is `Some(0)` it should stay `Some(0)`.

Reply:
Would it be better to store `next_is_elem as usize` as a variable? Does it affect compilation?

Reply:
Fixed